Attempt to create messages DB and watcher to track messages needing update between restarts

Nyefari opened the pull request:
I don't know how to handle typecasts and nullabilities correctly :/

Nyefari committed:
Merge pull request #1 from martindevans/master

Merging in Martin's Changes

Nyefari committed:
Merge pull request #2 from martindevans/master

Merging Changes by Martin on main May 29th

Nyefari committed:
Attempt to create DB for messages and watch it for messages to update. I don't know how to deal with type casts and possible nullability :/

Nyefari committed:
Should have been in last commit

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+            }
+        }
+
+        public async Task TrackMessage(ulong ChannelID, ulong MessageID, ulong ChallengeID, uint MessageType)
```
Make `MessageType` an enum? This would make it clearer at the usage sites.

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+                        var chn = channel as ISocketMessageChannel;
+                        IUserMessage msg = chn.GetMessageAsync(entry.MessageID) as IUserMessage; 
+
+                        await msg.ModifyAsync(a => a.Embed = c.ToEmbed().Build());
```
Add some extra null checks in case channel or message is null:

```
var channel = _client.GetChannel(entry.ChannelID) as ISocketMessageChannel;
IUserMessage msg = chn?.GetMessageAsync(entry.MessageID) as IUserMessage;
if (msg != null)
    await msg.ModifyAsync(a => a.Embed = c.ToEmbed().Build());
```

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+        {
+            _cron.Schedule(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1), uint.MaxValue, async () => {
+                // Get current challenge
+                var c = await _challenges.GetCurrentChallenge();
```
You need to handle `c` potentially being null i.e. when no challenge is running.

Nyefari replied:
now handled by just having the job do nothing if it is null.

martindevans commented on YololCompetition/Services/Messages/IMessages.cs:
```diff
@@ -0,0 +1,30 @@
+{
+    public interface IMessages
+    {
+        Task TrackMessage(ulong ChannelID, ulong MessageID, ulong ChallengeID, uint MessageType);
```
As mentioned above, use an `enum` for `MessageType`

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+            DbCommand PrepareQuery(IDatabase database)
+            {
+                var cmd = _database.CreateCommand();
+                cmd.CommandText = "SELECT from 'Messages' WHERE 'ChallengID' = @challengeID";
```
Typo on `ChallengID`

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+                var c = await _challenges.GetCurrentChallenge();
+
+                var messages = GetCurrentMessages(c.Id);
+                await foreach (Message entry in messages) 
```
Maybe wrap every case in a `try/catch` block and log errors - to ensure the cron job never dies.

Nyefari replied:
I need help with this one. I don't know what I'm doing with error handling and there's a few places I would like to implement it.

martindevans replied:
At the moment Referee does not have a structured logging system in place, so if you want to log out an error just call `Console.WriteLine(blah);`.

If you're unfamiliar with `try/catch` you can wrap a block of code:

```
try
{
    SomeCodeThatMightThrow();
}
catch (Exception e)
{
    HandleError(e);
}
```

This stops the exception propagating upwards because it stops in the `catch` block.

So in this case, I'm just suggesting that you put everything inside the foreach loop in a `try/catch` in a block, that way if any case fails it won't terminate the loop and the entire cron job. You _could_ do slightly more advanced error handling than just logging it e.g. remove that particular entry from the DB or wait a short while and retry this case. Since all entries are wiped from this table every 5 days I think it's good enough just to log and move on.

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+                    if (entry.ChallengeID != c.Id) {continue;}
+
+                    switch (entry.MessageType) {
+                        default:
```
Handle the `default` case as an error. e.g. by logging an error, or throwing an exception (but make sure not to kill this cron!)

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+
+        public enum messageType : uint {Current, Leaderboard}
+        
+        public DbMessages(IDatabase database, ICron cron, IChallenges challenges, BaseSocketClient client)
```
You either need to take `DiscordSocketClient` or add `BaseSocketClient` into the DI container in `Program.cs` (at the moment only `DiscordSocketClient` is added iirc, it's probably better to add both)

Nyefari replied:
DiscordSocketClient has the methods I need so I don't see why both would be better. I've changed it to take DiscordSocketClient

martindevans replied:
Sorry I wasn't very clear about that. What I mean is in `Program.cs` this code:

```
di.AddSingleton(new DiscordSocketClient(new DiscordSocketConfig {
    AlwaysDownloadUsers = true,
    MessageCacheSize = 0,
}));
```

Binds `DiscordSocketClient` into the DI container. However `BaseSocketClient` is _not_ bound into the container anywhere.

So what I was saying is you can either change your constructor to request `DiscordSocketClient` (since that's available in DI). Or you could change `Program.cs` to bind _both_ `DiscordSocketClient` and `BaseSocketClient` to the same item in DI, and then your constructor can continue requesting `BaseSocketClient`.

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,106 @@
+
+        public void StartMessageWatch()
+        {
+            _cron.Schedule(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1), uint.MaxValue, async () => {
```
This job will not update embeds after the challenge ends (to add the `Challenge has ended` text).

Nyefari replied:
Extracted Final Update to a separate method (`FinalUpdateMessages()` )that is run by the thing that ends the current competition. This updates them one last time (should get the ended message since it will be after the time the challenge ends) and removes them from the db messages table.

Nyefari committed:
Second Attempt at adding messages DB. Do not approve PR. Still needs try/catch in the cron job and various error handling.

Added FinalUpdateMessages() method to messages which is run by the endcurrentchallenge and clears old challenge's messages from the table. Corrected various issues & typos noted by Martin in previous PR attempt to main repo.
